### PR TITLE
Destination-S3: Correctly generate int64 values

### LINKED ISSUE #17564: Destination S3: Correctly generate int64 values
As noted in https://github.com/airbytehq/airbyte/pull/14362#discussion_r928033781, dest-s3's avro and parquet output formats currently use int32. This causes problems for users with int64 values. We should have dest-s3 produce int64 values for `integer` inputs.

The reason we use int32 right now is because we want to support union types, which behave kind of weirdly in edge cases. Consider this JSON schema:
```json
{
  "oneOf": [
    {"type": "string", "airbyte_type": "timestamp_with_timezone"},
    {"type": "number", "airbyte_type": "integer"}
  ]
}
```

Ideally, this would resolve to an Avro union:
```json
[
  {"type": "long", "logicalType": "timestamp-micros"},
  "long"
]
```

The problem is that Avro will detect that both types are actually `long` under the hood, so this isn't actually a valid union type. Furthermore, the json<>avro converter behaves differently based on the logic type:
* `{"created_at": 1634982000}` can be parsed as `"long"`, but not as `{"type": "long", "logicalType": "timestamp-micros"}`
* `{"created_at": "1634982000"}` (note the quotes around `1634982000`) can be parsed as `{"type": "long", "logicalType": "timestamp-micros"}`, but not as `"long"`.

Options:
* Decide that the union-typing is too niche of a usecase (maybe true) and stop supporting it
* Otherwise, we need to:
    * Decide what the correct output type is (either long or timestamp)
    * Patch the converter library to parse values correctly, given that decision

### PR DESCRIPTION
**Pull request by etsybaev**
## What
As noted in [#14362 (comment)](https://github.com/airbytehq/airbyte/pull/14362#discussion_r928033781), dest-s3's avro and parquet output formats currently use int32. This causes problems for users with int64 values. We should have dest-s3 produce int64 values for integer inputs.

The reason we use int32 right now is because we want to support union types, which behave kind of weirdly in edge cases.

## How
Updated S3 to use Avro.LONG type instead of Avro.INT

## 🚨 User Impact 🚨
At the first glance, no breaking changes are expected. In the edge case user might need to reset connection and re-sync data.
**Important note: If we would get oneOf ("long" and "long, specific type: timestamp" - the value would be treated as a long**

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Review by grishick — APPROVED** [review]

File: airbyte-integrations/bases/base-java-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonToAvroSchemaConverter.java [hunk 1]
```diff
@@ -471,4 +471,38 @@ Schema parseJsonField(final String fieldName,
...
+    }
+
+    if (hasPlainLong && hasTimestampMicrosLong) {
+      newUnionTypes.removeIf(type -> type.getType() == Schema.Type.LONG
+          && type.getLogicalType() != null
+          && type.getLogicalType().getName().equals("timestamp-micros"));
+      return Schema.createUnion(newUnionTypes);
+    } else {
+      return Schema.createUnion(unionTypes);
+    }
```
Comment (jbfbell) [inline airbyte-integrations/bases/base-java-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonToAvroSchemaConverter.java:505]: Logic is fine, I got a little carried away, so heres an alternate approach
```
private Schema createUnionAndCheckLongTypesDuplications(List<Schema> unionTypes) {
    Predicate<Schema> isALong = type -> type.getType() == Schema.Type.LONG;
    Predicate<Schema> isPlainLong = isALong.and(type -> Objects.isNull(type.getLogicalType()));
    Predicate<Schema> isTimestampMicrosLong = isALong.and(type ->
        Objects.nonNull(type.getLogicalType()) && "timestamp-micros".equals(type.getLogicalType().getName()));

    boolean hasPlainLong = unionTypes.stream().anyMatch(isPlainLong);
    boolean hasTimestampMicrosLong = unionTypes.stream().anyMatch(isTimestampMicrosLong);
    Predicate<Schema> removeTimestampType = type -> !(hasPlainLong && hasTimestampMicrosLong && isTimestampMicrosLong.test(type));
    return Schema.createUnion(unionTypes.stream().filter(removeTimestampType).collect(Collectors.toList()));
 }
    ```

**Review by jbfbell — APPROVED** [review]

File: airbyte-integrations/bases/base-java-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonToAvroSchemaConverter.java (hunk 1 above)
Comment (etsybaev) [reply airbyte-integrations/bases/base-java-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonToAvroSchemaConverter.java]: Updated to use your approach. Thanks!

**Comment by etsybaev** [comment]
/publish connector=connectors/destination-bigquery-denormalized

**Pull request merged**